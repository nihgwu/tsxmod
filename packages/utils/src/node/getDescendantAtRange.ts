import type { Node, SourceFile } from 'ts-morph'
import { SyntaxKind } from 'ts-morph'
import { TreeMode, getChildrenFunction } from './getChildrenFunction'

/**
 * Get a descendant node between a start and end range.
 *
 * @citation
 * Forked from [ts-ast-viewer](https://github.com/dsherret/ts-ast-viewer/blob/main/site/src/compiler/getDescendantAtRange.ts)
 */
export function getDescendantAtRange(
  sourceFile: SourceFile,
  range: [number, number],
  mode: keyof typeof TreeMode = TreeMode.forEachChild
): Node {
  const getChildren = getChildrenFunction(mode)
  let bestMatch: { node: Node; start: number } = {
    node: sourceFile,
    start: sourceFile.getStart(),
  }

  searchDescendants(sourceFile)

  return bestMatch.node

  function searchDescendants(node: Node) {
    for (const child of getChildren(node)) {
      if (child.getKind() !== SyntaxKind.SyntaxList) {
        if (isBeforeRange(child.getEnd())) {
          continue
        }

        const childStart = child.getStart()

        if (isAfterRange(childStart)) {
          return
        }

        const isEndOfFileToken = child.getKind() === SyntaxKind.EndOfFileToken

        if (!isEndOfFileToken && !hasSameStart(child)) {
          bestMatch = { node: child, start: childStart }
        }
      }

      searchDescendants(child)
    }
  }

  function hasSameStart(node: Node) {
    return bestMatch.start === node.getStart() && range[0] === node.getStart()
  }

  function isBeforeRange(position: number) {
    return position < range[0]
  }

  function isAfterRange(nodeEnd: number) {
    return nodeEnd >= range[0] && nodeEnd > range[1]
  }
}
