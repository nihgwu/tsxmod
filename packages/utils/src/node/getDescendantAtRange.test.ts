import { Project } from 'ts-morph'
import { getDescendantAtRange } from './getDescendantAtRange'

const simpleSource = `
const a = 1
const b = 2
const c = 3
`.trim()

const complexSource = `
import React from 'react'

export function Button(props: {
    onClick: () => void
    children: React.ReactNode
}) {
    return <button onClick={props.onClick}>{props.children}</button>
}
`.trim()

test('gets nodes within an offset range', () => {
  const project = new Project()
  const simpleSourceFile = project.createSourceFile('simple.ts', simpleSource)
  const simpleNode = getDescendantAtRange(simpleSourceFile, [17, 18])

  expect(simpleNode.getText()).toMatchSnapshot()

  const complexSourceFile = project.createSourceFile(
    'complex.ts',
    complexSource
  )
  const functionText = 'function'
  const functionIndexStart = complexSource.indexOf(functionText)
  const functionIndexEnd = functionIndexStart + functionText.length - 1
  const complexNode = getDescendantAtRange(complexSourceFile, [
    functionIndexStart,
    functionIndexEnd,
  ])

  expect(complexNode.getText()).toMatchSnapshot()
})
