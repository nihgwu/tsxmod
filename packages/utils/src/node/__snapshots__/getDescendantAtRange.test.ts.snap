// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets nodes within an offset range 1`] = `"b"`;

exports[`gets nodes within an offset range 2`] = `
"export function Button(props: {
    onClick: () => void
    children: React.ReactNode
}) {
    return <button onClick={props.onClick}>{props.children}</button>
}"
`;
